sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -964,6 +965,155 @@ object CombineFilters extends Rule[LogicalPlan] with PredicateHelper {
+    else ex
+  }
+
+  /**
+   * Pulls up references to aliases from an earlier operator and replaces them with the
+   * raw expression they are associated with.
+   *
+   * The output name of the original expression is assumed to be the desired final name
+   * of the rewritten expression, so if necessary an alias is added to ensure the output
+   * name is correct.
+   *
+   * @param ex expression to re-write
+   * @param aliasMap aliases from the input operator, mapped to their expressions
+   * @return rewritten expression with intermediate aliases removed
+   */
+  def resolveAliasesMaintainingSchema(ex: NamedExpression,
+                                      aliasMap: AttributeMap[Expression]): NamedExpression = {
+    val ret = replaceAlias(ex, aliasMap)
+    ret match {
+      case namedEx: NamedExpression =>
+        if (namedEx.name != ex.name) {
+          Alias(ret, ex.name)(ex.exprId, ex.qualifier, Some(ex.metadata))
+        } else {
+          namedEx
+        }
+      case _ => Alias(ret, ex.name)(ex.exprId, ex.qualifier, Some(ex.metadata))
+    }
+  }
+
+  def collapseIntoOneAggregate(aggExprs: Seq[NamedExpression],
+                               groupExprs: Seq[Expression],
+                               childAgg: Aggregate): Aggregate = {
+
+    val aliasMap = AttributeMap(childAgg.aggregateExpressions.collect {
+      case a: Alias => (a.toAttribute, a.child)
+    })
+    val aliasResolvedAggExprs = aggExprs.map(resolveAliasesMaintainingSchema(_, aliasMap))
+    val aliasResolvedGroupExprs = groupExprs.map(ex => replaceAlias(ex, aliasMap))
+    Aggregate(aliasResolvedGroupExprs, aliasResolvedAggExprs, childAgg.child)
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    plan transform {
+      // The query execution/optimization does not guarantee the expressions are evaluated in order.
+      // We only can combine them if and only if both are deterministic.
+      case agg@Aggregate(groupExprs: Seq[Expression],
+                         projectionsOfAggregateNode: Seq[NamedExpression],
+                         childAgg@Aggregate(childGroupExprs, childAggExprs, grandChild)) =>
+
+      var collapsible = true
+      val collapsibleNestedAggs = projectionsOfAggregateNode.map(aggEx => {
+        // Don't need to rewrite the projected grouping key expressions, but want to maintain them
+        // in the list, so identify them early and keep them the same.
+        //
+        // The or condition is kind of a hack for "early exit" of this loop, so later iterations
+        // don't overwrite the value to again declare this collapsible
+        if (groupExprs.exists(ex => ex.semanticEquals(aggEx)) || !collapsible) aggEx
+        else {
+          aggEx match {
+            case a@Alias(outerAggExpr: AggregateExpression, _) =>
+              outerAggExpr.aggregateFunction match {
+                case _: Max | _: Min | _: Sum =>
+                  // look for the expressions in the inner aggregate that produce
```

Comment:
Looks like the cases can be optimized by this rule is limited:

1. Max, Min, Sum only
2. Only one inner aggregate is referred at outer aggregate
3. Outer uses subset grouping expressions of inner

Is it common to have the cases this rule can optimize?

Reply:
It does handle the outer aggregate referring to multiple inner aggregates, there are some tests included that show this.

The rule is somewhat limited, but these cases come up, especially when it isn't a single person writing the whole query. If you are consuming a view and adding to it, you can be interested in aggregating the same types of columns that the writer of the previous view did, but your view has been rolled up to some granularity, say by region and weekly bucketing, and for your use case you want to just aggregate to straight region level.

I am happy to try to expand the cases covered by the rule to make it more useful, this is just the set of cases I was able to figure out so far.